### PR TITLE
Rebalancing policy shouldn't prevent hard allocation decisions

### PR DESCRIPTION
**Pull request by bleskes**
#14259 added a check to honor rebalancing policies (i.e., rebalance only on green state) when moving shards due to changes in allocation filtering rules. The rebalancing policy is there to make sure that we don't try to even out the number of shards per node when we are still missing shards. However, it should not interfere with explicit user commands (allocation filtering) or things like the disk threshold wanting to move shards because of a node hitting the high water mark.
#14259 was done to address #14057 where people reported that using allocation filtering caused many shards to be moved at once. This is however a none issue - with 1.7 (where the issue was reported) and 2.x, we protect recovery source nodes by limitting the number of concurrent data streams they can open (i.e., we can have many recoveries, but they will be throttled). In 5.0 we came up with a simpler and more understandable approach where we have a hard limit on the number of outgoing recoveries per node (on top of the incoming recoveries we already had).

### CODE REVIEW

**Comment by bleskes** [comment]
@ywelsch can you take a look?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java [hunk 1]
```diff
@@ -557,8 +557,7 @@ private boolean moveShard(NodeSorter sorter, ShardRouting shardRouting, ModelNod
                 if (currentNode != sourceNode) {
                     RoutingNode target = currentNode.getRoutingNode();
                     Decision allocationDecision = allocation.deciders().canAllocate(shardRouting, target, allocation);
-                    Decision rebalanceDecision = allocation.deciders().canRebalance(shardRouting, allocation);
-                    if (allocationDecision.type() == Type.YES && rebalanceDecision.type() == Type.YES) { // TODO maybe we can respect throttling here too?
+                    if (allocationDecision.type() == Type.YES) { // TODO maybe we can respect throttling here too?
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: can we add a comment here why we don't check "canRebalance"? (maybe link to this PR?)

**Comment by ywelsch** [comment]
Left one comment to document decision. Feel free to push afterwards. LGTM. Thanks @bleskes

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java (hunk 1 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java]: sure.

**Pull request merged**